library/core/src/time.rs

```diff
@@ -1278,13 +1276,30 @@ macro_rules! try_from_secs {
         } else if exp < 0 {
             // the input is less than 1 second
             let t = <$double_ty>::from(mant) << ($offset + exp);
-            let nanos = (u128::from(NANOS_PER_SEC) * u128::from(t)) >> ($mant_bits + $offset);
-            (0, nanos as u32)
+            let nanos_offset = $mant_bits + $offset;
+            let nanos_tmp = u128::from(NANOS_PER_SEC) * u128::from(t);
+            let nanos = (nanos_tmp >> nanos_offset) as u32;
+            if nanos_tmp & (1 << (nanos_offset - 1)) == 0 {
+                (0, nanos)
+            } else if nanos + 1 != NANOS_PER_SEC {
+                (0, nanos + 1)
+            } else {
+                (1, 0)
+            }
         } else if exp < $mant_bits {
-            let secs = mant >> ($mant_bits - exp);
+            let secs = u64::from(mant >> ($mant_bits - exp));
             let t = <$double_ty>::from((mant << exp) & MANT_MASK);
-            let nanos = (<$double_ty>::from(NANOS_PER_SEC) * t) >> $mant_bits;
-            (u64::from(secs), nanos as u32)
+            let nanos_tmp = <$double_ty>::from(NANOS_PER_SEC) * t;
+            let nanos = (nanos_tmp >> $mant_bits) as u32;
+            if nanos_tmp & (1 << ($mant_bits - 1)) == 0 {
+                (secs, nanos)
+            } else if nanos + 1 != NANOS_PER_SEC {
```

Comment:
Does the tie mode really matter in our case? IIUC tie to even is the default rounding in IEEE 754 only because it results in a slower rate of error accumulation for repeated additions/subtractions, which clearly is irrelevant in our case.

Implementing tie to even in software will be a bit unpleasant:
```rust
let rem_mask = (1 << nanos_offset) - 1;
let rem_msb_mask = 1 << (nanos_offset - 1);
let rem = nanos_tmp & rem_mask;
let is_tie = rem == rem_msb_mask;
let is_even = (nanos & 1) == 0;
let rem_msb = nanos_tmp & rem_msb_mask == 0;
if rem_msb || (is_even && is_tie) {
    (0, nanos)
} else {
    if nanos + 1 != NANOS_PER_SEC {
        (0, nanos + 1)
    } else {
        (1, 0)
    }
}
```

Reply:
I think it is critical that we implement the same tie resolution mechanism as used by the regular floating operations if we are implementing rounding-to-nearest for this specific operation at all.

Yes, implementing tie-away-from-zero is simpler. It is also true that the specific points raised in the referenced issue will get resolved no matter which tie resolution scheme is adopted here. I acknowledge that there are very few examples where the need to break ties comes up in the first place, and that the motivations behind picking a rounding mode for floating point operations doesn't apply here to the same degree.

What I'm worried about, however, is that somebody eventually comes up with some specific input that doesn't look “right”, would work right with a different tie resolution mode, and declares it a regression. With ties-away-from-zero their case would be that much stronger precisely because this scheme is so obscure.

Basically ties-to-even has a benefit of making it work exactly like the rest of the floating point operations, and in turn much easier to justify the specific choice for this function. If it doesn’t for some reason work right for somebody, it is then up to themselves to implement something that they think works right.

---

All that said, I'm be happy to leave the final call up to libs-api or some other T-libs team here.